Sources/BSON/BSONDocument.swift

```diff
@@ -174,12 +186,107 @@ public struct BSONDocument {
      */
     public subscript(dynamicMember member: String) -> BSON? {
         get { self[member] }
-        set { fatalError("Unimplemented") }
+        set { self[member] = newValue }
+    }
+
+    /**
+     * Sets a BSON element with the corresponding key
+     * if element.value is nil the element is deleted from the BSON
+     */
+    internal mutating func set(key: String, of value: BSON?) throws {
+        if !self.keySet.contains(key) {
+            guard let value = value else {
+                // no-op: key does not exist and the value is nil
+                return
+            }
+            // appending new key
+            self.keySet.insert(key)
+            self._buffer.moveWriterIndex(to: self.byteLength - 1) // setup to overwrite null terminator
+            let size = BSONDocument.append(key: key, of: value, to: &self._buffer)
+            self._buffer.writeInteger(0, endianness: .little, as: UInt8.self) // add back in our null terminator
+            self.byteLength += size
+            return
+        }
+
+        var iter = BSONDocumentIterator(over: self._buffer)
+
+        guard let (start, end) = iter.findByteRange(for: key) else {
+            throw BSONError.InternalError(message: "Cannot find \(key) to delete")
+        }
+
+        var newBuffer = BSON_ALLOCATOR.buffer(capacity: 0)
+
+        guard
+            let prefix = self._buffer.getBytes(at: 0, length: start),
+            let suffix = self._buffer.getBytes(at: end, length: self.byteLength - end)
+        else {
+            throw BSONError.InternalError(
+                message: "Cannot slice buffer from " +
+                    "0 to len \(start) and from \(end) to len \(self.byteLength - end) : \(self._buffer)"
+            )
+        }
+
+        newBuffer.writeBytes(prefix)
+
+        var newSize = self.byteLength - (end - start)
+        if let value = value {
+            // Overwriting
+            let size = BSONDocument.append(key: key, of: value, to: &newBuffer)
+            newSize += size
+
+            guard newSize <= BSON_MAX_SIZE else {
+                throw BSONError.DocumentTooLargeError(value: value.bsonValue, forKey: key)
+            }
+        } else {
+            // Deleting
+            self.keySet.remove(key)
+        }
+
+        newBuffer.writeBytes(suffix)
+
+        self._buffer = newBuffer
+        self.byteLength = newSize
+        guard self.byteLength == self._buffer.readableBytes else {
+            fatalError("BSONDocument's encoded byte length is \(self.byteLength) however the" +
+                "buffer has \(self._buffer.readableBytes) readable bytes")
+        }
+    }
+
+    /// Appends element to underlying BSON bytes, returns the size of the element appended: type + key + value
+    @discardableResult
+    internal static func append(key: String, of value: BSON, to buffer: inout ByteBuffer) -> Int {
```

Comment:
Maybe the BSONDocument is just getting too large and I could pull this and set out into an extension but I worry about the storage idea being a frustrating debugging experience, I've seen this pattern done for ByteBuffer and Data where they use a "storage" dataclass to wrap a pointer it's just adding yet another layer for those who need to peer into their BSON binary

Reply:
>Maybe the BSONDocument is just getting too large and I could pull this and set out into an extension

It's not necessarily just about stuff on `BSONDocument` though. Some related functionality is also added as extensions to `ByteBuffer` in the later PRs, so the logic for operating on BSON types with `ByteBuffer` is spread around the codebase in ways that aren't consistently discoverable. By introducing a new type that operates over the buffer, we can organize all this functionality into a single, easily discoverable place. 

>I worry about the storage idea being a frustrating debugging experience, I've seen this pattern done for ByteBuffer and Data where they use a "storage" dataclass to wrap a pointer it's just adding yet another layer for those who need to peer into their BSON binary

As far as causing debugging frustration, I guess I'm not really seeing what the concern is. Keep in mind this `DocumentStorage` would be a `struct`, so it wouldn't change the reference counting / CoW semantics of `ByteBuffer`. For me personally, encapsulating all this behavior in a single place would make debugging easier, since I could more easily enumerate all the places that touch the buffer.

Reply:
Alright yea I see the benefit there, I'll move the BSON modifying functions append and set as well as validate to a storage type, but I have to leave the extensions to ByteBuffer there since they're needed in reading and writing types and the BSONValue protocol operates on ByteBuffer. 
I was too vague about the debug frustration 😅  it's literally in the xcode UI there'd be 2 more layers of drop downs to see the internals of a BSON document. But that's a very specific developer pain so not too relevant